Ensure the `read_to_end` buffer always has enough room to fit a single UTF-8 code point

ChrisDenton opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/142872)*
<!-- homu-ignore:end -->

This is a quick fix to partly address rust-lang/rust#142847. So long as the buffer we read into has space for at least one UTF-8 code point then we can avoid any issues caused by splitting between code point boundaries (thus avoiding a lot of complexity). I think this is also good in general as it avoids some unnecessarily short reads.

a1phyr commented on library/std/src/io/mod.rs:
```diff
@@ -464,7 +464,8 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
         }
 
-        if buf.len() == buf.capacity() {
+        // Ensure there's at least enough space for one UTF-8 encoded code point.
+        if buf.spare_capacity_mut().len() < char::MAX_LEN_UTF8 {
             // buf is full, need more space
             buf.try_reserve(PROBE_SIZE)?;
         }
```
I think that this the condition could even be replaced by an unconditional call to `try_reserve(PROBE_SIZE)`.

the8472 replied:
That might lead to some unnecessary reallocations in some edge-cases, see #89165 (also mentioned in the function's top comment). It's tricky to get these tradeoffs right.

Maybe we can limit this to cfg(windows)? Specialization could work too but feels a bit overkill here.

the8472 replied:
Also, please add the issue to the list of masters this function is serving. Anyone who wants to touch it will have to keep track of all these things.

the8472 commented:
Since `default_read_to_end` is a crate-private we could also add a "min read capacity" option which can then be passed down from the stdin impl

the8472 commented:
Hrrmm, actually... isn't Stdin buffered? Shouldn't that buffer prevent that tearing? I don't see how read_to_end bypasses that.

a1phyr commented on library/std/src/io/mod.rs:
```diff
@@ -452,7 +452,7 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
 
     loop {
-        if buf.len() == buf.capacity() && buf.capacity() == start_cap {
+        if buf.spare_capacity_mut().len() < PROBE_SIZE && buf.capacity() == start_cap {
```
I thought that it would be a good idea too, but `small_probe_read` is designed to check for EOF so taking this branch if we still have space in the vector does not seem right.

the8472 replied:
- initial cap == 0 → this branch won't be hit
- initial cap == exact size → we should be getting an Ok(0) here and return
- short reads → don't need a large buffer anyway, looping here is fine
- other cases → reading into the probe buffer will then grow the vec, which means we don't hit this branch again

Yeah this seems fine.

a1phyr commented on library/std/src/io/mod.rs:
```diff
@@ -462,9 +462,14 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
-        if buf.len() == buf.capacity() {
+        // Try to ensure we're not making unnecessarily short reads.
+        // This is skipped if the user's pre-allocated Vec may have been exactly sized.
+        if buf.capacity() != start_cap || buf.spare_capacity_mut().len() == 0 {
```
How does this new condition avoids the issue that's supposed to be fixed? With this code we can still do short reads. I still think that calling `try_reserve` without check for the spare capacity is the simplest, and keeping the check about `start_cap` avoids over-allocations.

ChrisDenton replied:
Hm, the only case where the spare capacity will be less than `PROBE_SIZE` after this point is when `small_probe_read` reallocates but the reallocation ends up smaller than `PROBE_SIZE`. Given `Vec`'s reallocation strategy. I'm not sure if that's an issue in practice given the loop above, I guess the `if` statement is indeed made redundant.

a1phyr reviewed (approved):
This version looks great! Thanks!

~Side note: I think that these lines are no longer relevant, now that `default_read_to_end` uses the unstable `read_buf` API.~

https://github.com/rust-lang/rust/blob/ae2fc9722f08ef131407c1dc8057768868f65e8e/library/std/src/io/mod.rs#L404-L405

ibraheemdev commented:
This makes sense to me but r? the8472 for the final approval.

a1phyr commented:
Friendly ping @the8472, could you merge this please?

the8472 commented on library/std/src/io/mod.rs:
```diff
@@ -405,7 +405,7 @@ where
 // - pass large buffers to readers that do not initialize the spare capacity. this can amortize per-call overheads
 // - and finally pass not-too-small and not-too-large buffers to Windows read APIs because they manage to suffer from both problems
```
nit: the "finally" is no longer appropriate here

the8472 commented on library/std/src/io/mod.rs:
```diff
@@ -462,12 +462,13 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
-            buf.try_reserve(PROBE_SIZE)?;
-        }
+        // Avoid unnecessarily short reads by ensuring there's at least PROBE_SIZE space available.
+        buf.try_reserve(PROBE_SIZE)?;
```
This could use a const assert that `PROBE_SIZE >=  MAX_LEN_UTF8` since that's what's actually fixing the issue.

Of course a probe smaller than that would seem silly, but I'm not misremembering we had a 1 byte probe in the past.

the8472 reviewed (commented):
I guess this isn't easy to test since it needs a real console producing inputs? Otherwise a test would be nice.

Since the opening comment says

> resolve https://github.com/rust-lang/rust/issues/142847.

You might want to check the issue again, a user there reported that the issue won't be fixed by this alone.

Beyond that, just some nits.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/142872/6b3ae3f6e45a33c2d95fa0362c9b2593e567fd34..5473402d84bd09d63c1f6ca48499b00f0d9833ee)*

ChrisDenton committed:
Avoid unnecessarily short reads in `read_to_end`

On Windows, the UTF-16 to UTF-8 translation is made simpler by ensuring we don't split code points.
